# Ocean Environmental Monitoring - Data Validation

This project develops an end-to-end data pipeline for monitoring ocean environmental conditions using satellite-derived datasets from Copernicus Marine Service.

The system integrates multiple oceanographic variables: Sea Surface Temperature (SST); Surface wind speed; Significant wave height

Data is extracted for strategically selected coastal locations along Brazil, covering different oceanographic regimes:
- Tropical (Salvador)
- Major port regions (Rio de Janeiro, Santos)
- High-energy southern coast (Florianópolis, Rio Grande)

The objective is to:
- Monitor environmental variability
- Identify spatial and temporal patterns
- Detect potentially critical ocean conditions

This notebook (**01_data_validation.ipynb**) focuses on validating the dataset, ensuring:
- Data integrity and completeness
- Consistent temporal coverage across variables
- Physically plausible value ranges

The analytical exploration and insights are developed in the subsequent notebook:
**02_ocean_analysis.ipynb**

In [ ]:
# Connects to PostgreSQL database

import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(
    "postgresql+psycopg2://USER:PASSWORD@localhost:5432/ocean_monitoring"
)

In [2]:
# SQL query to retrieve measurements joined with location metadata

query = """
SELECT 
    l.name AS location,
    l.latitude,
    l.longitude,
    m.datetime,
    m.sst_c,
    m.wind_speed_ms,
    m.wave_height_m
FROM measurements m
JOIN locations l ON l.id = m.location_id
"""

# Load data into a pandas DataFrame
df = pd.read_sql(query, engine)

# First rows visualization
df.head()

,location,latitude,longitude,datetime,sst_c,wind_speed_ms,wave_height_m
0,florianopolis,-27.6,-48.4,2024-01-01 00:00:00,24.074387,4.089214,1.39
1,florianopolis,-27.6,-48.4,2024-01-01 03:00:00,23.911112,1.930376,1.24
2,florianopolis,-27.6,-48.4,2024-01-01 06:00:00,23.755568,1.990851,1.18
3,florianopolis,-27.6,-48.4,2024-01-01 09:00:00,23.733880,1.334412,1.18
4,florianopolis,-27.6,-48.4,2024-01-01 12:00:00,23.911728,4.074447,1.12


In [3]:
# Check dataset dimensions (rows, columns)
df.shape

(29205, 7)

In [4]:
# Check for missing values in each column
df.isnull().sum()

location          0
latitude          0
longitude         0
datetime          0
sst_c             0
wind_speed_ms     0
wave_height_m    20
dtype: int64

In [5]:
# Remove rows with missing values, if detected above
# Ensures consistency across all variables for analysis
df = df.dropna()

# Inspect data types and verifies non-null counts
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 29185 entries, 0 to 29204
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   location       29185 non-null  object        
 1   latitude       29185 non-null  float64       
 2   longitude      29185 non-null  float64       
 3   datetime       29185 non-null  datetime64[ns]
 4   sst_c          29185 non-null  float64       
 5   wind_speed_ms  29185 non-null  float64       
 6   wave_height_m  29185 non-null  float64       
dtypes: datetime64[ns](1), float64(5), object(1)
memory usage: 1.8+ MB


In [6]:
# Ensure datetime column is properly parsed as datetime type
# Even if already correct, this guarantees consistency for time-based analysis

df["datetime"] = pd.to_datetime(df["datetime"])

# Identify temporal range of the dataset
df["datetime"].min(), df["datetime"].max()

(Timestamp('2024-01-01 00:00:00'), Timestamp('2025-12-31 00:00:00'))

In [7]:
# Distribution of records by year
df["datetime"].dt.year.value_counts().sort_index()

datetime
2024    14620
2025    14565
Name: count, dtype: int64

In [8]:
# List unique monitoring locations to confirm all expected locations are present
df["location"].unique()

array(['florianopolis', 'rio_de_janeiro', 'rio_grande', 'salvador',
       'santos'], dtype=object)

In [9]:
# Count number of records per location
df["location"].value_counts()

location
florianopolis     5837
rio_de_janeiro    5837
rio_grande        5837
salvador          5837
santos            5837
Name: count, dtype: int64

In [10]:
# Identify duplicate rows
# Duplicates may indicate ingestion or processing issues

df.duplicated().sum()

np.int64(0)

In [11]:
# Summary statistics for numerical variables (used to detect unrealistic values)
df[["sst_c", "wind_speed_ms", "wave_height_m"]].describe()

,sst_c,wind_speed_ms,wave_height_m
count,29185.000000,29185.000000,29185.000000
mean,23.499903,4.452159,1.322589
std,3.799358,2.619335,0.620657
min,12.454196,0.127131,0.240000
25%,21.066772,2.416516,0.900000
50%,23.892883,4.103237,1.220000
75%,26.417147,5.934402,1.630000
max,31.440378,19.295746,7.410000


In [12]:
# Compute average environmental conditions per location, providing an initial spatial comparison

df.groupby("location")[["sst_c", "wind_speed_ms", "wave_height_m"]].mean()

,sst_c,wind_speed_ms,wave_height_m
location,,,
florianopolis,22.223833,4.207687,1.204691
rio_de_janeiro,24.092261,2.372376,0.756423
rio_grande,19.603305,7.044451,1.975518
salvador,27.203635,5.283245,1.363771
santos,24.376482,3.353035,1.312541


### Data Validation Summary

- Temporal coverage spans from 2024-01-01 to 2025-12-31.
- The dataset contains 29205 total observations across 5 monitoring locations.
- After removing incomplete records, the dataset was reduced to 29185 observations with full variable coverage.
- The remaining data is consistent, with no duplicates and physically plausible values.

The dataset is considered valid and ready for analytical exploration.

## Next Step - Ocean Analysis

With the dataset validated and cleaned, the next step is to explore the data from an analytical perspective.

The following notebook, **02_ocean_analysis.ipynb**, focuses on:

- Exploratory data analysis (EDA)
- Temporal patterns and variability
- Spatial comparisons between locations
- Identification of extreme environmental conditions

This aims to transform the validated dataset into meaningful insights about ocean dynamics and environmental variability.